Feature/583 circular hierarchy

PROgram52bc opened the pull request:
- Utility functions and tests related to resolving the group leadership hierarchy
- Checks for circular structure when:
   - adding a manager/member
   - reactivating a manager/member
   - moving a manager/member
   - reactivating a group (not implemented)
- fixed some/most of the styling/linting errors in the group module

PROgram52bc committed:
added group hierarchy helper (undone)

PROgram52bc committed:
added treeview-hierarchy page, reorganize existing treeview page

PROgram52bc committed:
added hierarchy data generator

PROgram52bc committed:
update group hierarchy

- added unit test for utilities in GroupHierarchyNode
- added depth filtering to getAllSubNodes
- added cycle check to getTree

PROgram52bc committed:
update GroupHierarchyNode

- added map function to GraphNode to transform nodes
- added checkConnection function
- added tests for map and checkConnection function
- minor clean up

PROgram52bc committed:
update GroupHierarchyNode

- updated test data
- added minimal hierarchical tree
- added adminTree, using isRootNode function
- added test for isRootNode function

PROgram52bc committed:
updated GroupHierarchyNode

- made mapTree a stand-alone generic function
- removed 'transform' option in dfs (in favor of mapTree)
- minor changes

PROgram52bc committed:
update treeview component

- added style and other function to treeviewHierarchy
- minor change in category treeview component

PROgram52bc committed:
fix GroupParticipant bug

- fixed move multiple participant error
- fixed add multiple participant error
- closes #587 Add manager not working

PROgram52bc committed:
removed groupHierarchyMixin

- refactored mixin functionality into component and utility function

PROgram52bc committed:
update GroupTreeViewHierarchy

- added icons and links for nodes
- excluded inactive groups and participants in search
- made groups all active in test data

PROgram52bc committed:
update GroupParticipants page

- added cycling check on adding a member/manager
- renamed convertGroupMap to convertToGroupMap
- minor fix in HierarchyNode

PROgram52bc committed:
implemented cycle check in create/move/unarchive participants

PROgram52bc committed:
Merge branch 'fix/598-fix-vuetify-breaking-changes' into feature/583-circular-hierarchy

PROgram52bc committed:
Vuetify 2.0 related update

- re-added typescript dependency
- minor change in treeview and GroupParticipants
- reformatted files

PROgram52bc committed:
updated group hierarchy tests to include active attribute

PROgram52bc commented on ui/src/components/groups/GroupParticipants.vue:
```diff
@@ -582,7 +580,31 @@ export default {
           );
-        } // else if (moving participant will create cycle) {...}
+        } else {
+          // check whether moving moving participant will create cycle
```
typo

PROgram52bc commented on ui/src/models/GroupHierarchyNode.ts:
```diff
@@ -0,0 +1,488 @@
+/* checks whether a node is a root node
+ * a root node <r> is a node that either:
+ * 1. does not have any super-node, or
+ * 2. for all its immediate super-node <s> such that <r> is the only super-node of <s>
```
rephrase this

PROgram52bc committed:
fixed typo and phrasing errors

PROgram52bc committed:
removed unused variables and trailing space

PROgram52bc committed:
updated GroupParticipants

- avoid redundant fetch after add/edit/archive/unarchive participants

PROgram52bc committed:
updated GroupHierarchyNode

- added toHumanReadable method
- fixed cycle bug in getTree
- added optional 'path' parameter to HierarchyCycleError

PROgram52bc committed:
added cycle handling to GroupTreeViewHierarchy page

PROgram52bc committed:
fixed styling error in treeview

PROgram52bc committed:
added row click for GroupParticipants batch actions

PROgram52bc committed:
forceUpdate cyclingPersons in saveLocalParticipants

PROgram52bc committed:
fixed individual unarchiving bug

PROgram52bc committed:
parameterized active filter on HierarchyNode

nurkkala commented:
This is a big one, so I would rather review it when there are no conflicts.

Please resolve merge conflicts and resubmit.

PROgram52bc committed:
fixed saveLocalParticipant method in GroupParticipants

- now it updates both managers and members in each group/person
associated, regardless of whether isManagerMode, to update local data
thoroughly

PROgram52bc commented:
I will resolve it soon!

PROgram52bc commented on ui/src/components/groups/treeview/GroupTreeViewHierarchy.vue:
```diff
@@ -0,0 +1,205 @@
+            <template v-slot:activator="{ on }">
+              <v-icon v-on="on">error</v-icon>
+            </template>
+            Please remove the cycle in the node listed in order to view group
```
internationalize this

PROgram52bc committed:
update checkConnection method

- now allows a member to be both managers/members of multiple groups
- added test case
- reformat GroupHierarchyNode.spec.ts file

PROgram52bc committed:
Merge branch 'development' into feature/583-circular-hierarchy

PROgram52bc committed:
fixed EmailForm error

PROgram52bc committed:
fixed style warning in router.js

PROgram52bc committed:
changed all v-layout/v-flex to v-row/v-col in group module

PROgram52bc commented:
merged into #601

PROgram52bc commented:
merged into #601
